File: app/src/main/java/com/zinhao/kikoeru/AudioPlayerActivity.java
package com.zinhao.kikoeru;

import android.content.ComponentName;
import android.content.Intent;
import android.content.ServiceConnection;
import android.graphics.Bitmap;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.os.IBinder;
import android.support.v4.media.session.PlaybackStateCompat;
import android.view.Menu;
import android.view.MenuItem;
import android.view.SubMenu;
import android.view.View;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.palette.graphics.Palette;
import com.bumptech.glide.Glide;
import com.bumptech.glide.load.resource.bitmap.RoundedCorners;
import com.bumptech.glide.request.RequestOptions;
import com.bumptech.glide.request.target.CustomViewTarget;
import com.bumptech.glide.request.transition.Transition;
import com.google.android.exoplayer2.Player;
import com.koushikdutta.async.http.AsyncHttpClient;
import com.koushikdutta.async.http.AsyncHttpResponse;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Locale;

public class AudioPlayerActivity extends BaseActivity implements ServiceConnection, MusicChangeListener, LrcRowChangeListener, SeekBar.OnSeekBarChangeListener {
    private AudioService.CtrlBinder ctrlBinder;
    private ImageView imageView;
    private RequestOptions options;
    private ImageButton ibPrevious;
    private ImageButton ibPause;
    private ImageButton ibNext;
    private ImageButton ibLrc;
    private ImageButton ibLoop;
    private TextView tvLrc;
    private TextView tvUpLrc;
    private TextView tvNextLrc;
    private TimeProgressView timeProgressView;
    private boolean needShowLrcWhenDestroy = false;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_player);
        imageView = findViewById(R.id.ivCover);
        imageView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Api.doGetWork(String.valueOf(ctrlBinder.getCurrentAlbumId()), 1, searchWorkCallback);
            }
        });
        RoundedCorners roundedCorners = new RoundedCorners(20);
        options = RequestOptions.bitmapTransform(roundedCorners);
        ibPrevious = findViewById(R.id.ib1);
        ibPause = findViewById(R.id.ib2);
        ibNext = findViewById(R.id.ib3);
        ibLrc = findViewById(R.id.imageButton2);
        ibLoop = findViewById(R.id.ibLoop);
        tvLrc = findViewById(R.id.tvLrc);
        tvUpLrc = findViewById(R.id.tvUpLrc);
        tvNextLrc = findViewById(R.id.tvNextLrc);
        timeProgressView = findViewById(R.id.time_view);
        timeProgressView.setColor(ContextCompat.getColor(this, R.color.play_control_icon_color));
        ibPause.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (ctrlBinder == null)
                    return;
                if (ctrlBinder.getController() == null || ctrlBinder.getController().getTransportControls() == null)
                    return;
                PlaybackStateCompat playbackStateCompat = ctrlBinder.getController().getPlaybackState();
                if (playbackStateCompat != null && playbackStateCompat.getState() == PlaybackStateCompat.STATE_PLAYING) {
                    ctrlBinder.getController().getTransportControls().pause();
                } else {
                    ctrlBinder.getController().getTransportControls().play();
                }

            }
        });
        ibPrevious.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (ctrlBinder == null)
                    return;
                if (ctrlBinder.getController() == null || ctrlBinder.getController().getTransportControls() == null)
                    return;
                ctrlBinder.getController().getTransportControls().skipToPrevious();
            }
        });
        ibNext.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (ctrlBinder == null)
                    return;
                if (ctrlBinder.getController() == null || ctrlBinder.getController().getTransportControls() == null)
                    return;
                ctrlBinder.getController().getTransportControls().skipToNext();
            }
        });
        ibLoop.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (ctrlBinder == null)
                    return;
                if (ctrlBinder.getController() == null || ctrlBinder.getController().getTransportControls() == null)
                    return;
                if (ctrlBinder.getReapMode() == Player.REPEAT_MODE_ONE) {
                    ctrlBinder.setReapAll();
                } else if (ctrlBinder.getReapMode() == Player.REPEAT_MODE_ALL) {
                    ctrlBinder.setReapOff();
                } else if (ctrlBinder.getReapMode() == Player.REPEAT_MODE_OFF) {
                    ctrlBinder.setReapOne();
                }
                updateLoopIcon();
            }
        });
        timeProgressView.setOnSeekBarChangeListener(this);

        bindService(new Intent(this, AudioService.class), this, BIND_AUTO_CREATE);
    }

    private AsyncHttpClient.JSONObjectCallback searchWorkCallback = new AsyncHttpClient.JSONObjectCallback() {
        @Override
        public void onCompleted(Exception e, AsyncHttpResponse asyncHttpResponse, JSONObject jsonObject) {
            if (e != null) {
                alertException(e);
                return;
            }
            if (asyncHttpResponse.code() == 200) {
                try {
                    int totalCount = jsonObject.getJSONObject("pagination").getInt("totalCount");
                    if (totalCount < 1)
                        return;
                    JSONArray works = jsonObject.getJSONArray("works");
                    if (works.length() != 0) {
                        JSONObject item = works.getJSONObject(0);
                        Intent intent = new Intent(AudioPlayerActivity.this, WorkTreeActivity.class);
                        intent.putExtra("work_json_str", item.toString());
                        ActivityCompat.startActivity(AudioPlayerActivity.this, intent, null);
                    }
                } catch (JSONException jsonException) {
                    jsonException.printStackTrace();
                }

            }
        }
    };

    private void updateLoopIcon() {
        if (ctrlBinder == null)
            return;
        if (ctrlBinder.getReapMode() == Player.REPEAT_MODE_ONE) {
            ibLoop.setImageResource(R.drawable.ic_baseline_flip_camera_android_24);
        } else if (ctrlBinder.getReapMode() == Player.REPEAT_MODE_ALL) {
            ibLoop.setImageResource(R.drawable.ic_baseline_loop_24);
        } else if (ctrlBinder.getReapMode() == Player.REPEAT_MODE_OFF) {
            ibLoop.setImageResource(R.drawable.ic_baseline_close_24);
        }
    }

    @Override
    public void onServiceConnected(ComponentName name, IBinder service) {
        ctrlBinder = (AudioService.CtrlBinder) service;
        ctrlBinder.addMusicChangeListener(this);
        ctrlBinder.addLrcRowChangeListener(this);
        if (ctrlBinder.isLrcWindowShow()) {
            needShowLrcWhenDestroy = true;
            ctrlBinder.hideLrcFloatWindow();
        }
        updateLoopIcon();
        ibLrc.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if(ctrlBinder.isLrcWindowShow()){
                    ctrlBinder.hideLrcFloatWindow();
                }else{
                    ctrlBinder.showLrcFloatWindow();
                }
            }
        });
        timeProgressView.setMax((int) ctrlBinder.getExoPlayer().getDuration());
        updateSeek();
    }

    @Override
    public void onServiceDisconnected(ComponentName name) {
    }

    @Override
    public void onChange(Lrc.LrcRow currentRow) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                tvUpLrc.setText(currentRow.getUpRow().content);
                tvLrc.setText(currentRow.content);
                tvNextLrc.setText(currentRow.getNextRow().content);
            }
        });
    }

    @Override
    public void onAlbumChange(int rjNumber) {
        Glide.with(this).asBitmap().load(Api.formatGetUrl(String.format(Locale.US, "/api/cover/%d", rjNumber), true)).apply(options).into(new CustomViewTarget<ImageView, Bitmap>(imageView) {

            @Override
            public void onLoadFailed(@Nullable Drawable drawable) {

            }

            @Override
            public void onResourceReady(@NonNull Bitmap bitmap, @Nullable Transition<? super Bitmap> transition) {
                imageView.setImageBitmap(bitmap);
                Palette.from(bitmap).generate(new Palette.PaletteAsyncListener() {
                    @Override
                    public void onGenerated(@Nullable Palette palette) {
                        View bg = (View) imageView.getParent();
                        if(bg!=null&& palette!=null){
                            int mainColor = palette.getDarkMutedColor(ActivityCompat.getColor(AudioPlayerActivity.this,R.color.main_color));
                            bg.setBackgroundColor(mainColor);
                            getWindow().setNavigationBarColor(mainColor);
                            getWindow().setStatusBarColor(mainColor);
                            ActionBar actionBar = getSupportActionBar();
                            if(actionBar!=null)
                                actionBar.setBackgroundDrawable(new ColorDrawable(mainColor));

                        }

                    }
                });
            }

            @Override
            protected void onResourceCleared(@Nullable Drawable drawable) {

            }
        });
    }

    @Override
    public void onAudioChange(JSONObject audio) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                try {
                    timeProgressView.setMax((int) ctrlBinder.getExoPlayer().getDuration());
                    setTitle(ctrlBinder.getCurrentTitle());
                } catch (JSONException e) {
                    e.printStackTrace();
                    alertException(e);
                }
            }
        });
    }

    @Override
    public void onStatusChange(int status) {
        timeProgressView.setMax((int) ctrlBinder.getExoPlayer().getDuration());
        if (status == 0) {
            ibPause.setImageResource(R.drawable.ic_baseline_play_arrow_24);
        } else {
            ibPause.setImageResource(R.drawable.ic_baseline_pause_24);
        }
    }

    private void updateSeek() {
        timeProgressView.postDelayed(new Runnable() {
            @Override
            public void run() {
                if (ctrlBinder != null && ctrlBinder.getExoPlayer() != null && ctrlBinder.getExoPlayer().isPlaying()) {
                    long current = ctrlBinder.getExoPlayer().getCurrentPosition();
                    timeProgressView.setProgress((int) current);
                }
                if (!isDestroyed())
                    updateSeek();
            }
        }, 500);
    }

    @Override
    protected void onDestroy() {
        if (ctrlBinder != null) {
            ctrlBinder.removeLrcRowChangeListener(this);
            ctrlBinder.removeMusicChangeListener(this);
            unbindService(this);
            if (!ctrlBinder.isLrcWindowShow() && needShowLrcWhenDestroy)
                ctrlBinder.showLrcFloatWindow();
        }
        super.onDestroy();
    }

    @Override
    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
        if (fromUser) {
            ctrlBinder.getController().getTransportControls().seekTo(progress);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        SubMenu subMenu = menu.addSubMenu(0, 0, 0, "stop after");
        subMenu.add(1, 1, 1, "30 minutes");
        subMenu.add(1, 2, 2, "60 minutes");
        subMenu.add(1, 3, 3, "90 minutes");
        subMenu.add(1, 4, 4, "120 minutes");
        subMenu.add(1, 5, 5, "240 minutes");
        subMenu.getItem().setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        try {
            if (item.getItemId() == 1) {
                ctrlBinder.stopAfterMinutes(30);
            } else if (item.getItemId() == 2) {
                ctrlBinder.stopAfterMinutes(60);
            } else if (item.getItemId() == 3) {
                ctrlBinder.stopAfterMinutes(90);
            } else if (item.getItemId() == 4) {
                ctrlBinder.stopAfterMinutes(120);
            } else if (item.getItemId() == 5) {
                ctrlBinder.stopAfterMinutes(240);
            }
        } catch (Exception e) {

        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {

    }

    @Override
    public void onStopTrackingTouch(SeekBar seekBar) {

    }
}

File: app/src/main/java/com/zinhao/kikoeru/AudioPlayerActivity.kt
package com.zinhao.kikoeru

import android.content.ComponentName
import android.content.Intent
import android.content.ServiceConnection
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.os.IBinder
import android.support.v4.media.session.PlaybackStateCompat
import android.util.DisplayMetrics
import android.util.TypedValue
import android.view.View
import android.widget.*
import android.widget.SeekBar.OnSeekBarChangeListener
import androidx.appcompat.widget.ListPopupWindow
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.core.graphics.drawable.toDrawable
import androidx.palette.graphics.Palette
import androidx.palette.graphics.Palette.PaletteAsyncListener
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.request.target.CustomViewTarget
import com.bumptech.glide.request.transition.Transition
import com.google.android.exoplayer2.Player
import com.koushikdutta.async.http.AsyncHttpClient.JSONObjectCallback
import com.koushikdutta.async.http.AsyncHttpResponse
import com.zinhao.kikoeru.Api.doGetWork
import com.zinhao.kikoeru.Api.formatGetUrl
import com.zinhao.kikoeru.AudioService.CtrlBinder
import com.zinhao.kikoeru.Lrc.LrcRow
import com.zinhao.kikoeru.databinding.ActivityPlayerBinding
import org.json.JSONException
import org.json.JSONObject
import java.util.*

class AudioPlayerActivity : BaseActivity(), ServiceConnection, MusicChangeListener, LrcRowChangeListener,
    OnSeekBarChangeListener {
    private var ctrlBinder: CtrlBinder? = null
    private var imageView: ImageView? = null
    private lateinit var rootView: View
    private lateinit var tvTitle: TextView
    private var ibPrevious: ImageButton? = null
    private var ibPause: ImageButton? = null
    private var ibNext: ImageButton? = null
    private var ibSleep: ImageButton? = null
    private var ibLoop: ImageButton? = null
    private var recyclerView: RecyclerView? = null
    private var timeProgressView: TimeProgressView? = null
    private var needShowLrcWhenDestroy = false
    private var lrcAdapter: LrcAdapter? = null
    private lateinit var viewBinding: ActivityPlayerBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewBinding = ActivityPlayerBinding.inflate(layoutInflater)
        setContentView(viewBinding.root)
        setSafeArea(viewBinding.root)
        imageView = viewBinding.ivCover
        imageView!!.setOnClickListener { doGetWork(ctrlBinder!!.currentAlbumId.toString(), 1, searchWorkCallback) }
        ibPrevious =viewBinding.ib1
        ibPause = viewBinding.ib2
        ibNext = viewBinding.ib3
        ibSleep = viewBinding.imageButton2
        ibLoop = viewBinding.ibLoop
        tvTitle = viewBinding.textView13
        setupSleepMenu()
        timeProgressView = viewBinding.timeView
        timeProgressView!!.setColor(ContextCompat.getColor(this, R.color.play_control_icon_color))
        ibPause!!.setOnClickListener(object : View.OnClickListener {
            override fun onClick(v: View?) {
                if (ctrlBinder == null) return
                if (ctrlBinder!!.getController() == null || ctrlBinder!!.getController()
                        .getTransportControls() == null
                ) return
                val playbackStateCompat = ctrlBinder!!.controller.getPlaybackState()
                if (playbackStateCompat != null && playbackStateCompat.getState() == PlaybackStateCompat.STATE_PLAYING) {
                    ctrlBinder!!.controller.getTransportControls().pause()
                } else {
                    ctrlBinder!!.getController().getTransportControls().play()
                }
            }
        })
        ibPrevious!!.setOnClickListener(object : View.OnClickListener {
            override fun onClick(v: View?) {
                if (ctrlBinder == null) return
                if (ctrlBinder!!.getController() == null || ctrlBinder!!.getController()
                        .getTransportControls() == null
                ) return
                ctrlBinder!!.getController().getTransportControls().skipToPrevious()
            }
        })
        ibNext!!.setOnClickListener(object : View.OnClickListener {
            override fun onClick(v: View?) {
                if (ctrlBinder == null) return
                if (ctrlBinder!!.getController() == null || ctrlBinder!!.getController()
                        .getTransportControls() == null
                ) return
                ctrlBinder!!.getController().getTransportControls().skipToNext()
            }
        })
        ibLoop!!.setOnClickListener(object : View.OnClickListener {
            override fun onClick(v: View?) {
                if (ctrlBinder == null) return
                if (ctrlBinder!!.getController() == null || ctrlBinder!!.getController()
                        .getTransportControls() == null
                ) return
                if (ctrlBinder!!.reapMode == Player.REPEAT_MODE_ONE) {
                    ctrlBinder!!.setReapAll()
                    Toast.makeText(this@AudioPlayerActivity, getString(R.string.all_repeat), Toast.LENGTH_SHORT).show()
                } else if (ctrlBinder!!.reapMode == Player.REPEAT_MODE_ALL) {
                    ctrlBinder!!.setReapOff()
                    Toast.makeText(this@AudioPlayerActivity, getString(R.string.repeat_off), Toast.LENGTH_SHORT).show()
                } else if (ctrlBinder!!.reapMode == Player.REPEAT_MODE_OFF) {
                    ctrlBinder!!.setReapOne()
                    Toast.makeText(this@AudioPlayerActivity, getString(R.string.repeat_one), Toast.LENGTH_SHORT).show()
                }
                updateLoopIcon()
            }
        })
        timeProgressView!!.setOnSeekBarChangeListener(this)

        bindService(Intent(this, AudioService::class.java), this, BIND_AUTO_CREATE)
    }

    private var sleepMenu: ListPopupWindow? = null
    private fun setupSleepMenu() {
        sleepMenu = ListPopupWindow(this)
        sleepMenu!!.setAdapter(
            ArrayAdapter<String?>(
                this, android.R.layout.simple_list_item_1,
                listOf<String?>(
                    "30 minutes", "1 hours", "90 minutes", "2 hours", "150 minutes", "3 hours",
                )
            )
        )
        sleepMenu!!.setModal(true)
        sleepMenu!!.setAnchorView( ibSleep )
        sleepMenu?.width = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 130f, resources.displayMetrics).toInt()
        sleepMenu!!.setOnItemClickListener(AdapterView.OnItemClickListener { parent: AdapterView<*>?, view: View?, position: Int, id: Long ->
            sleepMenu!!.dismiss()
            val minutes= 30*(position+1)
            if(BuildConfig.DEBUG){
                ctrlBinder?.stopAfterMinutes(minutes/30)
            }else{
                ctrlBinder?.stopAfterMinutes(minutes)
            }
            Toast.makeText(this, "will stop after ${minutes} minutes", Toast.LENGTH_LONG).show()
        })
    }

    var lastScrollIDLE = 0L

    private fun setupLrc() {
        recyclerView = viewBinding.mainRecycler
        ctrlBinder?.lrc?.let {
            lrcAdapter = LrcAdapter(it)
            lrcAdapter?.setOnToHereClickListener { v->
                v?.tag?.let { tag->
                    if(tag is LrcRow){
                        ctrlBinder?.controller?.transportControls?.seekTo(tag.time)
                        lrcAdapter?.notifyDataSetChanged()
                    }
                }
            }
            recyclerView!!.layoutManager = LinearLayoutManager(this)
            recyclerView!!.adapter = lrcAdapter
            scrollToLrcPosition()

            recyclerView!!.addOnScrollListener(object : RecyclerView.OnScrollListener() {
                override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                    if(newState == RecyclerView.SCROLL_STATE_IDLE) {
                        lastScrollIDLE = System.currentTimeMillis()
                    }
                }
            })
        }
    }

    private val searchWorkCallback: JSONObjectCallback = object : JSONObjectCallback() {
        override fun onCompleted(e: Exception?, asyncHttpResponse: AsyncHttpResponse, jsonObject: JSONObject) {
            if (e != null) {
                alertException(e)
                return
            }
            if (asyncHttpResponse.code() == 200) {
                try {
                    val totalCount = jsonObject.getJSONObject("pagination").getInt("totalCount")
                    if (totalCount < 1) return
                    val works = jsonObject.getJSONArray("works")
                    if (works.length() != 0) {
                        val item = works.getJSONObject(0)
                        val intent = Intent(this@AudioPlayerActivity, WorkTreeActivity::class.java)
                        intent.putExtra("work_json_str", item.toString())
                        this@AudioPlayerActivity.startActivity(intent, null)
                    }
                } catch (jsonException: JSONException) {
                    jsonException.printStackTrace()
                }
            }
        }
    }

    private fun updateLoopIcon() {
        if (ctrlBinder == null) return
        if (ctrlBinder!!.getReapMode() == Player.REPEAT_MODE_ONE) {
            ibLoop!!.setImageResource(R.drawable.ic_baseline_flip_camera_android_24)
        } else if (ctrlBinder!!.getReapMode() == Player.REPEAT_MODE_ALL) {
            ibLoop!!.setImageResource(R.drawable.ic_baseline_loop_24)
        } else if (ctrlBinder!!.getReapMode() == Player.REPEAT_MODE_OFF) {
            ibLoop!!.setImageResource(R.drawable.ic_baseline_close_24)
        }
    }

    override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
        ctrlBinder = service as CtrlBinder?
        ctrlBinder!!.addMusicChangeListener(this)
        ctrlBinder!!.addLrcChangeListener(this)
        if (ctrlBinder!!.isLrcWindowShow()) {
            needShowLrcWhenDestroy = true
            ctrlBinder!!.hideLrcFloatWindow()
        }
        updateLoopIcon()
        ibSleep!!.setOnClickListener(object : View.OnClickListener {
            override fun onClick(v: View?) {
                sleepMenu?.show()
            }
        })
        timeProgressView!!.setMax(ctrlBinder!!.getExoPlayer().getDuration().toInt())
        if (ctrlBinder != null && ctrlBinder!!.getExoPlayer() != null) {
            val current = ctrlBinder!!.getExoPlayer().getCurrentPosition()
            val buffer = ctrlBinder!!.getExoPlayer().getBufferedPosition()
            timeProgressView!!.setProgress(current.toInt(), buffer.toInt())
        }
        updateSeek()
        if(ctrlBinder?.lrc== Lrc.NONE){
            runOnUiThread { imageView?.alpha = 1f }
        }else{
            runOnUiThread { imageView?.alpha = 0.5f }
        }
        setupLrc()
    }

    override fun onServiceDisconnected(name: ComponentName?) {
    }

    override fun onSeekChange(lrcRow: LrcRow) {
        runOnUiThread {
            lrcAdapter?.update()
            scrollToLrcPosition()
        }
    }

    private fun scrollToLrcPosition(){
        if(System.currentTimeMillis() - lastScrollIDLE > 3000) {
            val manger = recyclerView?.layoutManager
            if(manger is LinearLayoutManager) {
                val index = manger.findLastCompletelyVisibleItemPosition()
                ctrlBinder?.let {
                    if(it.lrc.currentIndex != index) {
                        manger.scrollToPosition(it.lrc.currentIndex)
                    }
                }
            }
        }
    }

    override fun onLrcChange(lrc: Lrc?) {
        if(ctrlBinder?.lrc== Lrc.NONE){
            runOnUiThread { imageView?.alpha = 1f }
        }else{
            runOnUiThread { imageView?.alpha = 0.5f }
        }
        runOnUiThread {
            setupLrc()
        }
    }

    override fun onAlbumChange(rjNumber: Long) {
        Glide.with(this).asBitmap().load(formatGetUrl(String.format(Locale.US, "/api/cover/%d", rjNumber), true))
            .apply(App.getInstance().radius15Pic).into(object : CustomViewTarget<ImageView?, Bitmap?>(imageView!!) {
                override fun onLoadFailed(drawable: Drawable?) {
                }

                override fun onResourceReady(bitmap: Bitmap, transition: Transition<in Bitmap?>?) {
                    imageView!!.setImageBitmap(bitmap)
                    Palette.from(bitmap).generate(object : PaletteAsyncListener {
                        override fun onGenerated(palette: Palette?) {
                            if (palette != null) {
                                val mainColor = palette.getDarkMutedColor(
                                    ActivityCompat.getColor(
                                        this@AudioPlayerActivity,
                                        R.color.main_color
                                    )
                                )
                                rootView.setBackgroundColor(mainColor)
                                window.setNavigationBarColor(mainColor)
                                window.setStatusBarColor(mainColor)
                                val actionBar = getSupportActionBar()
                                actionBar?.setBackgroundDrawable(mainColor.toDrawable())
                            }
                        }
                    })
                }

                override fun onResourceCleared(drawable: Drawable?) {
                }
            })
    }

    override fun onAudioChange(audio: JSONObject?) {
        runOnUiThread(object : Runnable {
            override fun run() {
                try {
                    timeProgressView!!.setMax(ctrlBinder!!.getExoPlayer().getDuration().toInt())
                    setTitle(ctrlBinder!!.getCurrentTitle())
                } catch (e: JSONException) {
                    e.printStackTrace()
                    alertException(e)
                }
            }
        })
    }

    override fun setTitle(title: CharSequence?) {
        tvTitle.text = title
    }

    override fun onStatusChange(status: Int) {
        timeProgressView!!.setMax(ctrlBinder!!.getExoPlayer().getDuration().toInt())
        if (status == 0) {
            ibPause!!.setImageResource(R.drawable.ic_baseline_play_arrow_24)
        } else {
            ibPause!!.setImageResource(R.drawable.ic_baseline_pause_24)
        }
    }

    private fun updateSeek() {
        timeProgressView!!.postDelayed(object : Runnable {
            override fun run() {
                if (ctrlBinder != null && ctrlBinder!!.getExoPlayer() != null && ctrlBinder!!.getExoPlayer()
                        .isPlaying()
                ) {
                    val current = ctrlBinder!!.getExoPlayer().getCurrentPosition()
                    val buffer = ctrlBinder!!.getExoPlayer().getBufferedPosition()
                    timeProgressView!!.setProgress(current.toInt(), buffer.toInt())
                }
                if (!isDestroyed()) updateSeek()
            }
        }, 500)
    }

    override fun onDestroy() {
        if (ctrlBinder != null) {
            ctrlBinder!!.removeLrcChangeListener(this)
            ctrlBinder!!.removeMusicChangeListener(this)
            unbindService(this)
            if (!ctrlBinder!!.isLrcWindowShow() && needShowLrcWhenDestroy) ctrlBinder!!.showLrcFloatWindow()
        }
        super.onDestroy()
    }

    override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) {
        if (fromUser) {
            ctrlBinder!!.getController().getTransportControls().seekTo(progress.toLong())
        }
    }
    override fun onStartTrackingTouch(seekBar: SeekBar?) {
    }

    override fun onStopTrackingTouch(seekBar: SeekBar?) {
    }
}
